### ISSUE TITLE
auto legend position changes upon saving the figure

### ISSUE THREAD

**Issue by anntzer** [matplotlib/matplotlib#7353]
2.0b4 and master

```
from matplotlib import pyplot as plt
import numpy as np

plt.rcdefaults()
np.random.seed(1)

plt.plot(*np.random.rand(2, 300), ".", label="some long legend")
plt.plot(*np.random.rand(2, 300), ".", label="other long legend")
plt.legend()
plt.savefig("/tmp/foo.png")
plt.show()
```

The displayed window puts the legend at the center top of the figure, but the saved image puts the legend at the bottom left.

By changing the seed one can get cases where the two legends do match each other.  In that case, careful inspection of the images show that the legends have slightly different sizes, which make them overlap with different number of points depending on the corner in which they are tentatively placed by the autolocator -- this is what leads to them moving upon saving.

**Comment by tacaswell** [comment]
The first time it saves:

![foo](https://cloud.githubusercontent.com/assets/199813/19712652/b378ed3c-9b0d-11e6-97b4-ebcd33576010.png)

Saving _after_ letting it render to the screen

![foo2](https://cloud.githubusercontent.com/assets/199813/19712660/d158e3ca-9b0d-11e6-9a79-f80806e5718c.png)

And they are slightly different sizes 

```python
In [63]: mimg.imread('/tmp/foo.png').shape
Out[63]: 
(480, 640, 4)

In [64]: mimg.imread('/tmp/foo2.png').shape
Out[64]: 
(471, 640, 4)
```

Which leads to the even weirder observation that `fig.set_size_inches` does not always take and some times comes back a bit shorter, which is in turn tied to `forward=True` as the new default.   If you set the same size _twice_ it takes.

```python
In [108]: plt.gcf().set_size_inches([6.4, 4.7], forward=True); print(plt.gcf().bbox_inches); print(plt.gcf().get_size_inches())
Bbox(x0=0.0, y0=0.0, x1=6.4, y1=4.61)
[ 6.4   4.61]

In [109]: plt.gcf().set_size_inches([6.4, 4.7], forward=True); print(plt.gcf().bbox_inches); print(plt.gcf().get_size_inches())
Bbox(x0=0.0, y0=0.0, x1=6.4, y1=4.7)
[ 6.4  4.7]
```

This seems to be related to something going wrong with the conversion between the size of the canvas and the size of the qt window.

**Comment by tacaswell** [comment]
closed by #7610
